src/metrics/emd/script.py

```diff
@@ -14,94 +13,113 @@
-    # harcoded the column file_name and use it in assert.
-    # See line 176 of https://github.com/TarikExner/CytoNormPy/blob/main/cytonormpy/_evaluation/_emd_utils.py#L173
-    input_donor.obs.rename(columns={'sample': 'file_name'}, inplace=True)
-
-    emd_integrated = cnp.emd_from_anndata(
-        adata=input_donor,
-        file_list=list(input_donor.obs['file_name'].unique()),
-        channels=markers_to_assess,
-        layer="data",
-        sample_identifier_column="file_name",
-        cell_labels="cell_type"
-    )
+# del input_unintegrated
+
+# get all donors invalidation as these are the ones we need to validate
+donor_list = input_validation.obs['donor'].unique()
+
+emd_per_donor_per_ct = []
+emd_per_donor_all_ct = []
+
+for donor in donor_list:
+    # donor = donor_list[0]
+    
+    integrated_view = input_integrated[input_integrated.obs['donor'] == donor]
+    validation_view = input_validation[input_validation.obs['donor'] == donor]
     
-    emd_integrated['donor'] = donor
-    emd_df.append(emd_integrated.copy())
+    # assuming each cell type is present in both validation and integrated
+    cell_types = validation_view.obs['cell_type'].unique()
     
-    # ignoring unlabeled and all_cells
-    emd_integrated.drop(index=['all_cells', 'Unlabeled'], inplace=True)
+    for cell_type in cell_types:
+        # cell_type = cell_types[0]
+        
+        integrated_ct = integrated_view[integrated_view.obs['cell_type'] == cell_type]
+        validation_ct = validation_view[validation_view.obs['cell_type'] == cell_type]
+        
+        # Do not calculate if we have less than 50 cells as it does not make sense.
+        if integrated_ct.n_obs < 50 or validation_ct.n_obs < 50:
+            continue
+        
+        emd_df = compute_emd(
+            integrated_ct = integrated_ct, 
+            validation_ct = validation_ct, 
+            markers_to_assess = markers_to_assess
+        )
+        emd_df['cell_type'] = cell_type
+        emd_df['donor'] = donor
+        
+        emd_per_donor_per_ct.append(emd_df)
+    
+    # calculate EMD when combining all cell types as well.
+    emd_df = compute_emd(
+        integrated_ct = integrated_view, 
+        validation_ct = validation_view, 
+        markers_to_assess = markers_to_assess
+    )
+    emd_df['cell_type'] = 'all_cell_types'
+    emd_df['donor'] = donor
     
-    emd_vals.extend(emd_integrated[markers_to_assess].to_numpy().flatten())
+    emd_per_donor_all_ct.append(emd_df)
     
-emd_df = pd.concat(emd_df)
+emd_per_donor_per_ct = pd.concat(emd_per_donor_per_ct)
+emd_per_donor_all_ct = pd.concat(emd_per_donor_all_ct)
+
+
+emd_mean_ct = np.nanmean(emd_per_donor_per_ct.drop(columns=['cell_type', 'donor']).values)
+emd_max_ct = np.nanmax(emd_per_donor_per_ct.drop(columns=['cell_type', 'donor']).values)
+
+emd_mean_dn = np.nanmean(emd_per_donor_all_ct.drop(columns=['cell_type', 'donor']).values)
```

Comment:
So the mean across donors is essentially the emd for all cell types then averaged across donors. I figured this will give a different emd information from emd_mean_ct in that emd_mean_ct is mean emd across the different cell types regardless of donors. While emd_mean_dn is mean emd across different donors regardless of cell types. I think it is a bit too complicated and we may have to think whether it makes sense to split it this way.